### ISSUE TITLE
[P0] Supplied FP32 diagonals pass the invertibility guard before FP16 serialization changes their meaning

### ISSUE THREAD

**Issue by RobTand** [RobTand/tessera#286]
Reviewed commit 4e9744f6424ecd0d90ebcbd4200c909123f38a77. This is a necessary follow-up to closed #229 / PR #267, not a repetition of its original fitted-factor witness. The repaired fit returns validated FP16 tensors and addresses that witness. The newly shared guard still accepts caller-supplied factors whose eventual stored words are invalid or different.

require_invertible_diagonals at src/tessera/diagonals.py:185-186 evaluates factor.float(), so a positive finite FP32 tensor passes regardless of its FP16 representation. encode_unit accepts these Diagonals and uses their FP32 values in apply_diagonals/undo_diagonals. The writer repeats the same guard at unit_artifact.py:613, then pack_fp16 at wire.py:315 silently casts the factors to FP16.

Consequences on the current source:
- Supplied sv=100000 (FP32) passes encoding and writing, but stores infinity. Supplied sv=1e-8 stores zero. In each case build_unit_artifact returns a 938-byte artifact that parse_unit_artifact rejects with a DIAG_SV noninvertibility error. The producer therefore successfully writes an artifact its reader cannot load.
- Supplied sv=1.0004 stays positive but rounds from 1.0003999471664429 to 1.0 on the wire. The artifact is readable, yet its reconstructed weights differ from the encoded unit used for pricing/verification. In the deterministic witness, maximum absolute reconstruction difference is 2.8118491172790527e-05. The guard must enforce the value used to encode as well as post-cast finiteness.

The supplied-factor API has Tensor fields, no FP16-only check, and no canonicalization before use. The new tests test_supplied_non_invertible_factors_are_refused and test_the_writer_refuses_non_invertible_diagonals in tests/test_diagonals.py:160-199 only provide already-FP16 invalid tensors, so they never exercise the later cast.

Cheap reproducible trigger, run from the reviewed checkout with PYTHONPATH=src OMP_NUM_THREADS=1 MKL_NUM_THREADS=1 OPENBLAS_NUM_THREADS=1 python3:

```python
import torch
from tessera.alphabet import E2M1_GRID, build_forest
from tessera.encode import encode_unit
from tessera.decode import reconstruct_unit
from tessera.diagonals import Diagonals
from tessera.trellis import ConvCode
from tessera.unit_artifact import build_unit_artifact, parse_unit_artifact

forest = build_forest(3, grid=E2M1_GRID)
code = ConvCode(memory=6)
weight = torch.randn(16, 32, generator=torch.Generator().manual_seed(7)) * .02
for factor in (100000., 1e-8, 1.0004):
    supplied = Diagonals(torch.full((16,), factor), torch.ones(32))  # FP32
    unit = encode_unit(weight, forest, (3,) * 32, code,
                       diagonals=supplied, scale_refit=1)
    before = reconstruct_unit(unit, forest, code)
    _, _, blob = build_unit_artifact(
        unit, "u", {3: forest}, 768, code, fixture_id=None)
    print(factor, len(blob))
    try:
        parsed = parse_unit_artifact(blob)
        after = reconstruct_unit(parsed.unit, forest, code)
        print(unit.diagonals.sv[0].item(), parsed.unit.diagonals.sv[0].item(),
              (before-after).abs().max().item())
    except Exception as error:
        print(type(error).__name__, error)
```

Recorded proof: /home/rob/tessera-runs/audit-recheck-2026-09-05/reviews/encoder_supplied_precision_probe.py and encoder_supplied_precision_probe.log. It exited zero on torch 2.10.0+cpu, device CPU, native math threads one. No model, GPU, PrismaBuild, or test suite was run.

Fix direction: define one canonical stored representation for supplied diagonals before any encoding, metric transport, or reconstruction uses them. Either require already-FP16 words explicitly, or land/cast once, validate the resulting words, and make every consumer use those same canonical values. The writer must fail before returning bytes if an EncodedUnit does not carry representable canonical factors; checking only the pre-cast FP32 tensor is insufficient.

Acceptance should cover FP32 and FP64 supplied factors that overflow, underflow, or merely round in FP16, for both DIAG_SV and DIAG_SU. Require a named rejection or canonical stored values consistently used by encode/transport/write/read; compare reconstructed weights before and after serialization for accepted values. Keep the original fitted-extreme and invalid-FP16 regressions.

No fix was attempted because the user explicitly requested review and necessary new issues without implementation changes. This proves a wire/representation defect; no served-quality measurement is claimed.

Disposition: filed during the requested review of closed audit issues and current code. No implementation changes were made; fixes are left for a separate dispatch.
